controller/tests/controller_server/utils_test.py

```diff
@@ -22,32 +23,118 @@ def setUp(self):
+
+    def _test_validate_secrets_with_config_valid_uid(self, uid):
+        secrets = {"config": str({uid: {"username": user, "password": password, "management_address": array}}
+                                 )}
+        utils.validate_secrets(secrets)
+
+    def test_validate_secrets_with_config_valid_uid(self):
+        self._test_validate_secrets_with_config_valid_uid("u i_d.")
+        self._test_validate_secrets_with_config_valid_uid("a" * 90)
+
+    def test_validate_secrets_with_config_and_topologies(self):
+        secrets = {"config": str({"u1": {"username": user, "password": password, "management_address": array,
+                                         "supported_topologies": [{"topology.kubernetes.io/test": "zone1",
+                                                                   "topology.block.csi.ibm.com/test": "dev1"}]}}
+                                 )}
+        utils.validate_secrets(secrets)
+
+    def _test_validate_secrets_with_config_invalid_parameters(self, uid="u1", topology="topology.kubernetes.io/test",
+                                                              value="test"):
+        secrets = {"config": str({uid: {"username": user, "password": password, "management_address": array,
+                                        "supported_topologies": [{topology: value}]}})}
+        self._test_validate_secrets_validation_errors(secrets)
+
+    def test_validate_secrets_with_config_invalid_parameters(self):
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u-1")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u:1")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u1+")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u1*")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u-1(")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u/1")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u=1")
+        self._test_validate_secrets_with_config_invalid_parameters(uid=" ")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="")
+        self._test_validate_secrets_with_config_invalid_parameters(uid=None)
+        self._test_validate_secrets_with_config_invalid_parameters(uid="a" * 91)
+        self._test_validate_secrets_with_config_invalid_parameters(topology="topology.kubernetes.io")
+        self._test_validate_secrets_with_config_invalid_parameters(topology="topology.kubernetes/test")
+        self._test_validate_secrets_with_config_invalid_parameters(topology="topology.kubernetes.io/t-est")
+        self._test_validate_secrets_with_config_invalid_parameters(value="a*")
+
+    def _test_get_pool_from_parameters(self, parameters, pool_expected=pool, uid=None):
+        pool_respond = utils.get_pool_from_parameters(parameters, uid)
+        self.assertEqual(pool_respond, pool_expected)
```

Comment:
consider renaming to `actual_pool` and `expected_pool` where applicable

Reply:
done